onedal/backend/data/data.cpp

```diff
@@ -0,0 +1,168 @@
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <stdexcept>
+#include <string>
+
+#include <Python.h>
+
+#include "backend/data/data.h"
+
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/row_accessor.hpp"
+#include "oneapi/dal/table/detail/homogen_utils.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+#include <CL/sycl.hpp>
+#endif
+
+#ifdef DPCTL_ENABLE
+#include "dpctl_sycl_types.h"
+#include "dpctl_sycl_queue_manager.h"
+#endif
+
+namespace oneapi::dal::python {
+#define is_array(a)         ((a) && PyArray_Check(a))
+#define array_type(a)       PyArray_TYPE((PyArrayObject *)a)
+#define array_is_behaved(a) (PyArray_ISCARRAY_RO((PyArrayObject *)a) && array_type(a) < NPY_OBJECT)
+#define array_is_behaved_F(a) \
+    (PyArray_ISFARRAY_RO((PyArrayObject *)a) && array_type(a) < NPY_OBJECT)
+#define array_is_native(a) (PyArray_ISNOTSWAPPED((PyArrayObject *)a))
+#define array_numdims(a)   PyArray_NDIM((PyArrayObject *)a)
+#define array_data(a)      PyArray_DATA((PyArrayObject *)a)
+#define array_size(a, i)   PyArray_DIM((PyArrayObject *)a, i)
+
+#ifdef ONEDAL_DATA_PARALLEL
+int init_numpy() {
+    import_array();
+    return 0;
+}
+
+const static int numpy_initialized = init_numpy();
+#endif
+
+template <typename T>
+inline dal::homogen_table convert_from_numpy_to_homogen(PyArrayObject *array) {
+    std::int64_t column_count = 1;
+
+    if (array_numdims(array) > 2) {
+        throw std::runtime_error("Input array has wrong dimensionality (must be 2d).");
+    }
+    T *data_pointer = reinterpret_cast<T *>(array_data(array));
+    // TODO: check safe cast from int to std::int64_t
+    const std::int64_t row_count = static_cast<std::int64_t>(array_size(array, 0));
+    if (array_numdims(array) == 2) {
+        // TODO: check safe cast from int to std::int64_t
+        column_count = static_cast<std::int64_t>(array_size(array, 1));
+    }
+    const auto layout =
+        array_is_behaved_F(array) ? dal::data_layout::column_major : dal::data_layout::row_major;
+    auto res_table = dal::homogen_table(data_pointer,
+                                        row_count,
+                                        column_count,
+                                        dal::detail::empty_delete<const T>(),
+                                        layout);
+    // we need it increment the ref-count if we use the input array in-place
+    // if we copied/converted it we already own our own reference
+    if (reinterpret_cast<PyArrayObject *>(data_pointer) == array)
+        Py_INCREF(array);
+    return res_table;
+}
+
+dal::table convert_from_numpy_to_table(PyObject *obj) {

```

Comment:
If it is constrained to numpy arrays only, it must accept array-like object as its parameter.

Reply:
not only a numpy pointer can come here